### PR TITLE
Add support for TRIAS Endpoints

### PR DESCRIPTION
**Pull request by 1Maxnet1**
This Pull Request introduces a draft for the required properties for trias clients and how to specify them. The information was collected from the existing list available in [trias-client](https://github.com/andaryjo/trias-client/blob/main/docs/PROVIDERS.md). I already found some issues, I would like to discuss:

- Some Endpoints offer test-endpoints, which are usually the same as the production endpoint but at least in one case they also [differ](https://github.com/andaryjo/trias-client/blob/main/docs/PROVIDERS.md) in Authorization (see the SBB). So what I iagined would be to use a boolean that indicate, whether this is a productive or test endpoint instead of offering the test-URL. This would lead to a lot of endpoints be defined two times, with only a different URL, but would allow to specify test-endpoints as that differ in more then the base URL.
- I wasn't sure how to define access: In the mentioned documentation sometimes a contract is required, but independent from whether a registration is required or not. Currently this is implicitly defined, if you need an authorization you usually need an registration as well. Except for the SBB testpoint, where you need an authorization with an publicly available key.
- A link to the documentation of the endpoint (and for registration) would be useful for developers. However this repo states to be "machine-readable" and in the other definitions, it is also not available. So should we add it or not?

Every kind of feedback is appreciated.

### CODE REVIEW

**Comment by derhuerst** [comment]
Thanks for your contribution!

My personal opinions on your points:

> * Some Endpoints offer test-endpoints, which are usually the same as the production endpoint but at least in one case they also [differ](https://github.com/andaryjo/trias-client/blob/main/docs/PROVIDERS.md) in Authorization (see the SBB). So what I iagined would be to use a boolean that indicate, whether this is a productive or test endpoint instead of offering the test-URL. This would lead to a lot of endpoints be defined two times, with only a different URL, but would allow to specify test-endpoints as that differ in more then the base URL.

I'm also in favour of having test endpoints as separate entries with e.g `isTestingEndpoint: true`.

> * I wasn't sure how to define access: In the mentioned documentation sometimes a contract is required, but independent from whether a registration is required or not. Currently this is implicitly defined, if you need an authorization you usually need an registration as well. Except for the SBB testpoint, where you need an authorization with an publicly available key.

I propose to keep technical means of authentication/authorization (e.g. keys) in a different field than legal/organizational means of authorization, mainly because the former is a clear cut, whereas the latter is a "fuzzy" topic.

> * A link to the documentation of the endpoint (and for registration) would be useful for developers. However this repo states to be "machine-readable" and in the other definitions, it is also not available. So should we add it or not?

I think having a link to a documentation page (or even multiple) is very helpful!

With this collection being machine-readable, we have the advantage of being able to filter or transform it into multiple representations (e.g. an HTML page equivalent, or reference is in code like `kpublictransport` & `hafas-client` do it); That does not mean that information intended for humans isn't helpful in here!

File: readme.md [hunk 1]
```diff
@@ -269,6 +269,20 @@ The following properties are defined:
...
+#### TRIAS
+
+*WIP*
+
+The following properties are defined:
+* `endpoint`: Base URL for the API
+* `testingEndpoint`: Base URL for testing purposes
+* `apiVersion`: `1.X` as [available](https://github.com/VDVde/TRIAS/releases)
+* `content-type`: [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) e.g. `text/xml`
+* `openAPI`: Boolean, whether one need to sign up or not
```
Comment (derhuerst) [inline readme.md]: I would rename this flag, given that [*OpenAPI*](https://www.openapis.org) is a very well-known spec for describing APIs in a machine-readable way. I propose this:

```suggestion
* `requiresSignUp`: Whether one needs to sign up to obtain an API key.
* `requiresContract` Wether one needs to sign a contract to use the API.
```

File: readme.md [hunk 2]
```diff
@@ -269,6 +269,20 @@ The following properties are defined:
...
 * `apiVersion`: One of `otp1`, `otp2` or `entur`.
 
+#### TRIAS
+
+*WIP*
+
+The following properties are defined:
+* `endpoint`: Base URL for the API
+* `testingEndpoint`: Base URL for testing purposes
+* `apiVersion`: `1.X` as [available](https://github.com/VDVde/TRIAS/releases)
```
Comment (derhuerst) [inline readme.md]:
```suggestion
* `triasVersion`: Highest [TRIAS version](https://github.com/VDVde/TRIAS/releases) supported by the API
```

File: readme.md [hunk 3]
```diff
@@ -269,6 +269,20 @@ The following properties are defined:
...
+* `content-type`: [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) e.g. `text/xml`
```
Comment (derhuerst) [inline readme.md]: Is this about the request or response?

File: readme.md [hunk 4]
```diff
@@ -269,6 +269,20 @@ The following properties are defined:
...
+* `openAPI`: Boolean, whether one need to sign up or not
+* `authorizationMethod`: One of `Authorization-Header`
+* `providedServices`: array of supported services as listed [in the documentation](https://www.vdv.de/431-2sdes-v-1-3.pdfx#page=51)
+* `requestLimit`: requests per day, which are allowed
```
Comment (derhuerst) [inline readme.md]: Shall we follow the [markup of the `RateLimit` HTTP header](https://www.ietf.org/archive/id/draft-ietf-httpapi-ratelimit-headers-04.html#name-expressing-rate-limit-polic) here? It is an established way of specifying the rate limit(s) of an API.

File: readme.md [hunk 5]
```diff
@@ -269,6 +269,20 @@ The following properties are defined:
...
+*WIP*
+
+The following properties are defined:
+* `endpoint`: Base URL for the API
+* `testingEndpoint`: Base URL for testing purposes
+* `apiVersion`: `1.X` as [available](https://github.com/VDVde/TRIAS/releases)
+* `content-type`: [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) e.g. `text/xml`
+* `openAPI`: Boolean, whether one need to sign up or not
+* `authorizationMethod`: One of `Authorization-Header`
+* `providedServices`: array of supported services as listed [in the documentation](https://www.vdv.de/431-2sdes-v-1-3.pdfx#page=51)
```
Comment (derhuerst) [inline readme.md]:
```suggestion
* `providedServices`: array of supported services as listed [in the EKAP documentation](https://www.vdv.de/431-2sdes-v-1-3.pdfx#page=51)
```

File: readme.md (hunk 3 above)
Comment (1Maxnet1) [reply readme.md]: This is about the request as some endpoints especially mention that this is required, see e.g. [here](https://www.vrn.de/opendata/node/115). I'm not sure about the response. Will check whether there are endpoints that differ in content-type of response and check whether there are clients, which differ in content-type of response and request. If the latter would exist, we would need two properties.

File: readme.md (hunk 4 above)
Comment (1Maxnet1) [reply readme.md]: This looks good, from a first view. Thanks! I will look into it and update the documentation accordingly.

File: readme.md (hunk 3 above)
Comment (derhuerst) [reply readme.md]: Even if the latter doesn't exist, I'd rename it to e.g. `request-content-type`.

**Comment by vkrause** [comment]
Thank you! This should also work as-is for OJP endpoints I think.

>   * Some Endpoints offer test-endpoints, which are usually the same as the production endpoint but at least in one case they also [differ](https://github.com/andaryjo/trias-client/blob/main/docs/PROVIDERS.md) in Authorization (see the SBB). So what I iagined would be to use a boolean that indicate, whether this is a productive or test endpoint instead of offering the test-URL. This would lead to a lot of endpoints be defined two times, with only a different URL, but would allow to specify test-endpoints as that differ in more then the base URL.

Including test endpoints in some form definitely makes sense, we also have that use-case for other protocols. Doing this as separate files is probably necessary though, due to differences in other properties like authentication, or e.g. Motis, which only has a testing endpoint so far.

>   * I wasn't sure how to define access: In the mentioned documentation sometimes a contract is required, but independent from whether a registration is required or not. Currently this is implicitly defined, if you need an authorization you usually need an registration as well. Except for the SBB testpoint, where you need an authorization with an publicly available key.
>
>   * A link to the documentation of the endpoint (and for registration) would be useful for developers. However this repo states to be "machine-readable" and in the other definitions, it is also not available. So should we add it or not?

"machine-readable" hits its limits once human interaction is required anyway, like for registering or negotiating access. Collecting that information here definitely makes sense, but it doesn't need to be machine readable IMHO.

How to best collect such service-specific notes came up before, one possibility would be to put that into a <servicename>.md file alongside the JSON data for example.

**Comment by 1Maxnet1** [comment]
> How to best collect such service-specific notes came up before, one possibility would be to put that into a .md file alongside the JSON data for example.

I would be fine with a .md file as well. What I came up with would be a property like `registrationURL` or sth. similar. How is this done for the hafas endpoints currently ? Is there any further information collected additionally to the attribution information? If not I would keep it that way for trias and would add a link to the site in `attritbution:{homepage: ...}`

**Comment by derhuerst** [comment]
> I would be fine with a .md file as well. What I came up with would be a property like `registrationURL` or sth. similar. How is this done for the hafas endpoints currently ?

The HAFAS endpoints that are in this repo right now don't require registration.

File: readme.md [hunk 6]
```diff
@@ -269,6 +269,21 @@ The following properties are defined:
...
 
+#### TRIAS
+
+*WIP*
+
+The following properties are defined:
+* `endpoint`: Base URL for the API
+* `testingEndpoint`: Base URL for testing purposes
+* `triasVersion`: Highest [TRIAS version](https://github.com/VDVde/TRIAS/releases) supported by the API
+* `request-content-type`: [Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type) e.g. `text/xml`
```
Comment (derhuerst) [inline readme.md]: Now this kebab-case style is different from `requestLimitPolicy`'s camelCase style. I suggest to change one of them.

**Comment by 1Maxnet1** [comment]
So I reworked it including your feedback and created a TRIAS-Providers.md, with an example entry. I thought adding the URL into the json wouldn't make to much sense, as technically it is not required for a client to use it. Let me know if you have any further points :+1:

**Review by derhuerst — APPROVED** [review]

**Comment by 1Maxnet1** [comment]
How do you handle commit history? In my opinion i could squash some of the commits together to keep the history cleaner. But I would be fine with leaving them as is as well. You can tell :D

**Comment by derhuerst** [comment]
> How do you handle commit history? In my opinion i could squash some of the commits together to keep the history cleaner.

We haven't discussed this.

In my opinion, when a PR is ready for merging, it makes sense to clean up the commit history to commits as atomic and orthogonal as (reasonably) possible. 👍

**Comment by 1Maxnet1** [comment]
> In my opinion, when a PR is ready for merging, it makes sense to clean up the commit history to commits as atomic and orthogonal as (reasonably) possible. +1

I cleaned it up :+1:  If there is anything you would like changed in the git history, just let me know. I also made a backup of the old one. Otherwise the PR looks good to me.

**Comment by 1Maxnet1** [comment]
I just created a .json file for the bwegt/NVBW TRIAS-Endpoint and recognized, they use another authorization method (https://github.com/1Maxnet1/transport-apis/tree/nvbw-endpoint). Should I add this addition to this PR or add it in another PR together with the submitted bwegt-trias.json ?

**Comment by derhuerst** [comment]
> I just created a .json file for the bwegt/NVBW TRIAS-Endpoint and recognized, they use another authorization method. Should I add this addition to this PR or add it in another PR together with the submitted bwegt-trias.json ?

What is different? Just https://github.com/1Maxnet1/transport-apis/commit/f5025fbabc3ee2bc499bb0016cf40789d0474ab2? Yeah, just add it here.

**Comment by 1Maxnet1** [comment]
> > I just created a .json file for the bwegt/NVBW TRIAS-Endpoint and recognized, they use another authorization method. Should I add this addition to this PR or add it in another PR together with the submitted bwegt-trias.json ?
> 
> What is different? Just [1Maxnet1@f5025fb](https://github.com/1Maxnet1/transport-apis/commit/f5025fbabc3ee2bc499bb0016cf40789d0474ab2)? Yeah, just add it here.

Yeah, just that, added it. Everything resolved now from my side

**Pull request merged**